text/2024-01-27-autohinting.md

```diff
@@ -0,0 +1,266 @@
+heuristic threshold) on a given axis. These are computed as might be expected:
+for each segment, we search for an existing edge whose position is within the
+distance threshold. If one isn't found, a new edge is created.
+
+The computed edge list is sorted by position: left to right for horizontal
+hinting and either top to bottom or bottom to top for vertical hinting,
+depending on the script assigned to the glyph-- most scripts hint from the
+baseline up but some scripts with a "hanging baseline" (such as the Indic
+group) produce better results when hinted top down.
+
+After edges are computed, the links between segments that were computed in
+the previous pass are propagated to the edges to form edge stems.
+
+Finally, we assign a blue (alignment) zone to each edge that falls within
+the region for that zone.
+
+### A simple visual example
+
+Consider the following lovely ASCII representation of a lowercase "t" where
+each point is signified by a letter (the extraneous G and J points added
+just to demonstrate segments with more than two points):
+<pre>
+                 A --- B
+                 |     |
+            M -- N     C -- D
+            |               |
+            L -- K     F -- E
+                 |     |
+                 J     G
+                 |     |
+                 I --- H
+</pre>
+
+Assuming hinting in the horizontal direction (left to right), we would compute
+the following topological features:
+
+* Segments (in outline order which is clockwise in this case):
+  * BC
+  * DE
+  * FGH
+  * IJK
+  * LM
+  * NA
+
+* Segment links (stems)
+  * NA <-> BC
+  * IJK <-> FGH
+
+> Note that DE and LM _do not_ form a link because the distance between them
+  surpasses the threshold.
+
+* Edges
+  * BC, FGH
+  * DE
+  * IJK, NA
+  * LM
+
+> Note that edges can be formed from non-contiguous segments.
+
+* Edge links (stems)
+  * BC, FGH <-> IJK, NA
+
+## Time for a change
+
+Once we've computed the topological features of our outline, we can start
+making adjustments. In general, grid-fitting is applied directly to our
+computed _edges_ and then those changes are propagated to the outline
+_points_.
+
+### Hinting edges ([src](https://github.com/googlefonts/fontations/blob/b283d608f7beca834c7d56cb649c79ba40c12411/skrifa/src/outline/autohint/hint/edges.rs))
+
+Edge hinting is performed in three passes:
+
+1. Edges that have been assigned a blue zone are handled first. These edges are
+   simply snapped to the grid-fitted position of their associated blue zone.
+2. Stem edges (those that have been linked with another edge to form a pair) are
+   handled next. These are snapped directly to the grid without guidance from an
+   alignment zone. The hinter tries to maintain a minimal distance from already
+   hinted edges to avoid overlaps.
+3. Remaining edges (single point edges and serifs) are handled last. These don't
+   form stems so are just fitted independently.
+
```

Comment:
Suggest referring to pair of edges in 1. and 2., I thought 1. was a single edge until I reached the paragraph below